### DISCUSSION TITLE
Issues replacing content of a component

### DISCUSSION THREAD

**Discussion by kangarko** [KyoriPowered/adventure#902] [General]
In calling TextComponent#content(String) it appears to ignore the new content.

Context: I am using ProtocolLib and listening to ClientboundSystemChatPacket which has two fields, content and adventure$content. My goal is to edit these fields before the packet gets sent.

Before adventure, this was simple. Now I am trying to get the adventure$content field which works, but after setting its new content it appears not to work anymore.

The relevant code excerpt is here. Please note that the system.out prints out "ClientboundSystemChatPacket -> set to: " as if the new content is still empty even after it has been set.

<img width="1051" alt="Screenshot 2023-04-12 at 18 53 19" src="https://user-images.githubusercontent.com/1918426/231528328-1585d196-8f02-403e-9936-ad039c558d4e.png">

Using Paper build 510.

**Comment by Machine-Maker** [comment]
Components are immutable. Calling `adventureContent.content(message)` doesn't mutate `adventureContent`, it just returns a new component with that content.

**Comment by kangarko** [comment]
So how can I modify the content of the ClientboundSystemChatPacket?

**Reply by kezz** [reply]
There are a few ways - you can use `Component#replaceText` or recreate the component entirely.

**Reply by kangarko** [reply]
I am sorry I am really not familiar with your library and would appreciate help. I tried calling replaceText to no success using recursion (because the packet "Unknown command. Type "/help" for help." message appears to be hidden in one of the children components):

<img width="576" alt="Screenshot 2023-04-12 at 19 12 57" src="https://user-images.githubusercontent.com/1918426/231532887-c13bc39d-5c4b-4de2-a825-5da2376a113e.png">

**Reply by kezz** [reply]
Again, components are immutable. This means those methods *return a new copy* of the component with the mutation applied.

**Reply by kangarko** [reply]
I see. My goal is to edit the field in the packet which is an Adventure's field. So there's no way to modify that? Thank you for the fast replies.

<img width="821" alt="Screenshot 2023-04-12 at 19 17 04" src="https://user-images.githubusercontent.com/1918426/231534095-aa4d8d03-a9d9-4846-8a08-65310e16be58.png">

**Reply by kezz** [reply]
Well, this is Adventure support not Java support. Changing a final field is up to you, all I can tell you is how components work.

**Reply by kangarko** [reply]
I understand and I appreciate the fast support. All I needed was a way to connect to Adventure using the packet which Paper has edited to use your library. I've opened a ticket at ProtocolLib. 

Afaik it looks unsolvable as long as either Paper doesn't remove the final modifier from the field or you don't add that change in your implementation.

**Reply by kangarko** [reply]
Discussed in https://github.com/PaperMC/Paper/issues/9128

**Reply by kezz** [reply]
We won't change anything our side - we already have mutable components in the form of component builders. Whatever other implementations/users of our library do is up to them.

**Reply by kangarko** [reply]
I understand, that is correct and I respect that is out of your bounds. Thanks so far.

**Comment by Greazi-Times** [comment]
Altho this librarry could be used. Protocollib has a working option.
The changes in 1.19 make it that you need to use Reflection. This isn't possible with java 8

There for there needs to be made a work around to get this working.
I have to take a look again how this needs to be done but this is the best shot on fixing this issue.

**Reply by zml2008** [reply]
> The changes in 1.19 make it that you need to use Reflection. This isn't possible with java 8

what?

1. reflection is nothing new
2. this doesn't stop you from doing a different strategy, like having your packet manipulation system replace one packet instance with another one that you construct yourself

this is a fairly common design pattern, there are loads of ways to handle changing one element nested in an immutable object graph.

**Reply by Greazi-Times** [reply]
Well we have a system that changes  messages send to a player.
With 1.19 it got changed and there for it din't work.

We have added a check for 1.19 and lower if thats true its using the old way.
But once its 1.19 or higher it is using a whole other system.

This systeem takes a lot of information about the message as thats now needed to make it work.

**Reply by kangarko** [reply]
Thanks guys, I was able to recreate the packet and fixed it! https://github.com/kangarko/Foundation/commit/8374209e590ce25a10888a861d046ecd7c6cc83c
